clang/lib/AST/ASTContext.cpp

```diff
@@ -3140,6 +3140,269 @@ ASTContext::getPointerAuthVTablePointerDiscriminator(const CXXRecordDecl *RD) {
   return llvm::getPointerAuthStableSipHash(Str);
 }
 
+/// Encode a function type for use in the discriminator of a function pointer
+/// type. We can't use the itanium scheme for this since C has quite permissive
+/// rules for type compatibility that we need to be compatible with.
+///
+/// Formally, this function associates every function pointer type T with an
+/// encoded string E(T). Let the equivalence relation T1 ~ T2 be defined as
+/// E(T1) == E(T2). E(T) is part of the ABI of values of type T. C type
+/// compatibility requires equivalent treatment under the ABI, so
+/// CCompatible(T1, T2) must imply E(T1) == E(T2), that is, CCompatible must be
+/// a subset of ~. Crucially, however, it must be a proper subset because
+/// CCompatible is not an equivalence relation: for example, int[] is compatible
+/// with both int[1] and int[2], but the latter are not compatible with each
+/// other. Therefore this encoding function must be careful to only distinguish
+/// types if there is no third type with which they are both required to be
+/// compatible.
+static void encodeTypeForFunctionPointerAuth(ASTContext &Ctx, raw_ostream &OS,
```

Comment:
In a subsequent patch, it's probably worth adding a (unit) test for encoding types as a string for ptrauth discrimination purposes. For this patch, clang/test/CodeGen/ptrauth-function-type-discriminator.c looks enough, but in future, a test which checks strings corresponding to types, and not final hash computed from strings, is nice to have.

Reply:
We can add use `__builtin_ptrauth_type_discriminator`, which hasn't been upstreamed yet, and `__builtin_ptrauth_string_discriminator` to test function type discriminators in a subsequent patch.

Reply:
This would be nice, thanks